nat-lab/guidelines.md

```diff
@@ -0,0 +1,212 @@
+
+   **Why:**:
+   - Stdout might say "Connected", but there are no actual connection in conntrack records.
+
+   **Example:**
+   - Use connection tracker. Providing configuration to environment setup will enable it. When test being teardown, it will be checked, wether actual connection to vpn server was made.
+   ```
+   def test_connection():
+      env = setup_env(SetupParams(connection_tag=ConnectionTag.DOCKER_CONE_CLIENT_1, connection_tracker_config=generate_connection_tracker_config(vpn_1_limits=ConnectionLimits(1, 1))))
+      client.connect_to_vpn()
+      client.wait_for_peer_state(Connected)
+      assert "Connected" in client.get_stdout()
+   ```
+
+8. **Comprehensive Node State Checking:**
+   - Examine the states of all related nodes, not just a single one.
+
+   **Why:**
+   - By checking only one node, it is not made sure, that other node will be the same.
+
+   **Example:**  
+   To check if two nodes are connected via meshnet PING is used.
+   - DO:
+   ```
+   def test_meshnet(alpha, beta):
+      env = setup_env([alpha, beta])
+
+      async with Ping(alpha, beta.ip).run() as ping:
+         await ping.wait_for_next_ping()
+      async with Ping(beta, alpha.ip).run() as ping:
+         await ping.wait_for_next_ping()
+
+      # assume that both peers are available
+   ```
+   - DON'T, because one of the node, might be blocking other node in firewall and by being able to ping from one side, you can't assume that connection is available both ways:
+   ```
+   def test_meshnet(alpha, beta):
+      env = setup_env([alpha, beta])
+
+      async with Ping(alpha, beta.ip).run() as ping:
+         await ping.wait_for_next_ping()
+
+      # assume that both peers are available
+   ```
+   because, this might have been the actual configuraton:
+   ```
+      alpha.set_peer_firewall_settings(beta, False)
+      beta.set_peer_firewall_settings(alpha, True)
+   ```
+
+9. **Replace Sleep with Event Checks:**
+   - Substitute sleep commands with event checks for enhanced reliability.
+
+   **Example:**  
+   Starting listening with `netcat`:
+   - DON'T:
+   ```
+   await connection_alpha.create_process("nc -nluv -4 100.64.0.100 12345")
+   await asyncio.sleep(2)
+   # assume that listening started
+   ```
+
+   - DO:
+   ```
+   output_notifier = OutputNotifier()
+   listening_start_event = asyncio.Event()
+   process = connection_alpha.create_process("nc -nluv -4 100.64.0.100 12345", on_stdout = output_notifier.handle_output())
+   output_notifier.notify_output(f"Bound on 100.64.0.100 12345", listening_start_event)
+   await process.execute()
+   await listening_start_event.wait()
+
+   # listening started for sure
+   ```
+
+10. **Prefer Wrapper Classes:**
+    - Opt for wrapper classes when interacting with external tools .
+
+    **Why:**
+    - Improved encapsulation
+    - Maintainability
+    - Uniformity
+    - Multiple OS support
```

Comment:
"tools."

Also, you mean here our wrappers, like `stun` one? Asking because python offers a lot of wrappers for different things as well, but they aren't particularly helpful when running things on the different containter.

Reply:
was talking about our own wrappers like ping.py; stun.py and etc. Updated doc